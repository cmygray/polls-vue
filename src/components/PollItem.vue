<template>
  <v-card tag="article" class="poll my-4">
    <v-card-title class="title">
      {{ poll.poll_title }}
    </v-card-title>

    <v-card-text>
      <dl v-if="isInProgress">
        <dt><small>시작</small></dt>
        <dd>{{ poll.starts_at | dateTime }}</dd>
        <dt><small>종료</small></dt>
        <dd v-if="isInProgress">
          {{ poll.ends_at | dateTime }}
        </dd>
      </dl>

      <p v-else>종료된 투표입니다</p>

      <v-radio-group
        v-if="isInProgress"
        v-model="selectedChoiceId"
        color="primary"
        label="Choices"
        hide-details
      >
        <div
          v-for="choice in choices"
          :key="`${choice.id}-${choice.votes_count}`"
        >
          <span
            ><small>득표수: {{ choice.votes_count }}</small></span
          >
          <v-radio
            :label="choice.choice_text"
            :value="choice.id"
            color="primary"
          />
        </div>
      </v-radio-group>
    </v-card-text>

    <v-card-actions>
      <v-btn
        v-if="isInProgress"
        :disabled="!selectedChoiceId"
        color="primary"
        block
        @click="handleClick"
        >Vote</v-btn
      >
    </v-card-actions>
  </v-card>
</template>

<script lang="ts">
import { Component, Emit, Prop, Vue } from 'vue-property-decorator';
import { Choice, Poll } from '@/models';

@Component
export default class PollItem extends Vue {
  @Prop()
  poll: Poll;

  selectedChoiceId: string = '';

  now: Date = new Date();

  get isInProgress(): boolean {
    return +new Date(this.poll.ends_at) > +this.now;
  }

  get choices(): Choice[] {
    return this.poll.choices.sort((current, next) => {
      return +current.id - +next.id;
    });
  }

  @Emit('poll-item:vote')
  handleClick() {
    return this.choices.find(choice => choice.id === this.selectedChoiceId);
  }
}
</script>
