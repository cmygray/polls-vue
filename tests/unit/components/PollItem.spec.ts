import { createLocalVue, mount } from '@vue/test-utils';

import PollItem from '@/components/PollItem.vue';
import { poll } from '../__fixtures__';
import '../setup';

const localVue = createLocalVue();

const NOW = new Date('2019-06-18T11:00:00.000Z')
const PAST = '2019-06-17T11:00:00.000Z'
const FUTURE = '2019-06-19T11:00:00.000Z'

describe('PollItem.vue', () => {
  let wrapper: any;

  describe('when ongoing poll is provided as prop', () => {
    beforeEach(() => {
      wrapper = mount(PollItem, {
        localVue,
        propsData: {
          poll: {
            ...poll,
            ends_at: FUTURE
          }
        },
        data: () => ({ now: NOW })
      });
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });
  })

  describe('when ended poll is provided as prop', () => {
    beforeEach(() => {
      wrapper = mount(PollItem, {
        localVue,
        propsData: {
          poll: {
            ...poll,
            ends_at: PAST
          }
        },
        data: () => ({ now: NOW })
      });
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });
  })

  describe('#handleClick', () => {
    const SELECTED_CHOICE = poll.choices[0];
    const EVENT = 'poll-item:vote'

    it('should find and emit selected choice with expected event', () => {
      wrapper.setData({
        selectedChoiceId: SELECTED_CHOICE.id
      })
      wrapper.vm.handleClick()

      expect(wrapper.emitted()[EVENT]).toEqual(expect.arrayContaining([
        [SELECTED_CHOICE]
      ]))
    })
  })
});
