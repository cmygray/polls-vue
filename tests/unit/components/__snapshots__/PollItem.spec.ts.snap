// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollItem.vue when ended poll is provided as prop should render as expected 1`] = `
<article class="poll my-4 v-card v-sheet theme--light">
  <div class="v-card__title title">
    TEST
  </div>
  <div class="v-card__text">
    <p>종료된 투표입니다</p>
    <!---->
  </div>
  <div class="v-card__actions">
    <!---->
  </div>
</article>
`;

exports[`PollItem.vue when ongoing poll is provided as prop should render as expected 1`] = `
<article class="poll my-4 v-card v-sheet theme--light">
  <div class="v-card__title title">
    TEST
  </div>
  <div class="v-card__text">
    <dl>
      <dt><small>시작</small></dt>
      <dd>2019-06-18 15:28</dd>
      <dt><small>종료</small></dt>
      <dd>
        2019-06-19 20:00
      </dd>
    </dl>
    <div class="v-input v-input--selection-controls v-input--radio-group v-input--radio-group--column v-input--hide-details theme--light">
      <div class="v-input__control">
        <div class="v-input__slot" style="height: auto;">
          <div role="radiogroup" class="v-input--radio-group__input"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">Choices</label>
            <div><span><small>득표수: 0</small></span>
              <div class="v-radio theme--light">
                <div class="v-input--selection-controls__input"><input aria-label="항목을 입력하세요" aria-checked="false" role="radio" type="radio" name="v-radio-4" value="5">
                  <div class="v-input--selection-controls__ripple"></div><i aria-hidden="true" class="v-icon material-icons theme--light">radio_button_unchecked</i>
                </div><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">항목을 입력하세요</label>
              </div>
            </div>
            <div><span><small>득표수: 0</small></span>
              <div class="v-radio theme--light">
                <div class="v-input--selection-controls__input"><input aria-label="항목을 입력하세요" aria-checked="false" role="radio" type="radio" name="v-radio-4" value="6">
                  <div class="v-input--selection-controls__ripple"></div><i aria-hidden="true" class="v-icon material-icons theme--light">radio_button_unchecked</i>
                </div><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">항목을 입력하세요</label>
              </div>
            </div>
            <div><span><small>득표수: 0</small></span>
              <div class="v-radio theme--light">
                <div class="v-input--selection-controls__input"><input aria-label="항목을 입력하세요" aria-checked="false" role="radio" type="radio" name="v-radio-4" value="7">
                  <div class="v-input--selection-controls__ripple"></div><i aria-hidden="true" class="v-icon material-icons theme--light">radio_button_unchecked</i>
                </div><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: relative;">항목을 입력하세요</label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-card__actions"><button disabled="disabled" type="button" class="v-btn v-btn--block v-btn--disabled theme--light primary--text">
      <div class="v-btn__content">Vote</div>
    </button></div>
</article>
`;
